lib/src/test/java/io/ably/lib/test/rest/RestRequestTest.java

```diff
@@ -0,0 +1,554 @@
+			assertTrue("Verify Content-Type header present", httpListener.getResponseHeader(req.id, "Content-Type").contains("application/json"));
+		} catch(AblyException e) {
+			e.printStackTrace();
+			fail("request_post: Unexpected exception");
+			return;
+		}
+	}
+
+	/**
+	 * Publish a message using the requestAsync() API
+	 */
+	@Test
+	public void request_post_async() {
+		final Waiter waiter = new Waiter();
+		final String messageData = "Test data (request_post_async)";
+		DebugOptions opts;
+		try {
+			opts = new DebugOptions(testVars.keys[0].keyStr);
+			fillInOptions(opts);
+			final RawHttpTracker httpListener = new RawHttpTracker();
+			opts.httpListener = httpListener;
+			AblyRest ably = new AblyRest(opts);
+
+			/* publish a message */
+			Message message = new Message("Test event", messageData);
+			JsonRequestBody requestBody = new JsonRequestBody(message);
+			ably.requestAsync(Http.POST, channelMessagesPath, null, requestBody, null, new Callback<AsyncPaginatedResult<JsonElement>>() {
+				@Override
+				public void onSuccess(AsyncPaginatedResult<JsonElement> result) {
+					/* wait to persist */
+					try { Thread.sleep(1000L); } catch(InterruptedException ie) {}
+
+					/* get the history */
+					Param[] params = new Param[] { new Param("limit", "1") };
+					PaginatedResult<Message> resultPage;
+					try {
+						resultPage = setupAbly.channels.get(channelName).history(params);
+
+						/* check it looks like a result page */
+						waiter.assertNotNull(resultPage);
+						waiter.assertTrue(resultPage.items().length == 1);
+						waiter.assertEquals(messageData, resultPage.items()[0].data);
+
+						/* check request has expected attributes */
+						RawHttpRequest req = httpListener.values().iterator().next();
+						/* Spec: RSC19b */
+						waiter.assertNotNull(httpListener.getRequestHeader(req.id, "Authorization"));
+						/* Spec: RSC19c */
+						waiter.assertTrue(httpListener.getRequestHeader(req.id, "Accept").contains("application/json"));
+						waiter.assertTrue(httpListener.getRequestHeader(req.id, "Content-Type").contains("application/json"));
+						waiter.assertTrue(httpListener.getResponseHeader(req.id, "Content-Type").contains("application/json"));
+						waiter.resume();
+					} catch (AblyException e) {
+						e.printStackTrace();
+						waiter.fail("request_post_async: Unexpected exception");
+						waiter.resume();
+					}
+				}
+				@Override
+				public void onError(ErrorInfo reason) {
+					waiter.fail("request_post_async: Unexpected exception");
+					waiter.resume();
+				}
+			});
+
+			try {
+				waiter.await(15000);
+			} catch (TimeoutException e) {
+				fail("request_paginated_async: Operation timed out");
+			}
+		} catch(AblyException e) {
+			e.printStackTrace();
+			fail("request_post_async: Unexpected exception");
+			return;
+		}
+	}
+
+	/**
+	 * Verify error responses are indicated with an exception
+	 * Spec: RSC19e
+	 */
+	@Test
+	public void request_404() {
```

Comment:
This is not correct according to the spec for two reasons:

* Exceptions are only raised when an HTTP response is not received, see http://docs.ably.io/client-lib-development-guide/features/#RSC19e
* Because you are not returning an `HttpPaginatedResponse`, you cannot check for a status of 404, or `success=false`.  I think unfortunately this PR does not address a lot of the requirements detailed in `HP1 - HP8` as a result, see http://docs.ably.io/client-lib-development-guide/features/#HP1